File: app/backend/app/api/ordenes.py
"""Endpoints de Ordenes de Reparacion.

Cada POST representa UNA intencion de una persona, no un nodo del
proceso. Una accion humana arrastra los nodos ACT-SYSTEM que la siguen
hasta la proxima frontera de actor humano; ahi se corta. Por eso no hay
un endpoint por nodo ni nada que ejecute el escenario completo de una.

El router es fino a proposito: valida la forma del request, delega en
``app.application`` y arma la respuesta. No decide nada de negocio y no
atrapa excepciones -de eso se ocupa ``errores.py``-.
"""

from fastapi import APIRouter, status

from app.application import (
    ApplicationContext,
    acciones_disponibles,
    aprobar_control,
    completar_ejecucion,
    crear_orden,
    definir_reparacion,
    encolar_orden,
    entregar,
    iniciar_detalle,
    insumos_previstos_por_detalle,
    listar_ordenes,
    nombres_de_tipo_por_detalle,
    notificar,
    obtener_orden,
    progreso,
    registrar_pago_de_orden,
    tomar_orden_en_estacion,
)
from app.domain.models import OrdenReparacion

from .dependencias import ContextoDep
from .schemas import (
    AprobarControlIn,
    CompletarEjecucionIn,
    CrearOrdenIn,
    DefinirReparacionIn,
    EncolarIn,
    EntregarIn,
    IniciarDetalleIn,
    NotificarIn,
    OrdenConEntregaOut,
    OrdenOut,
    OrdenResumenOut,
    PagoIn,
    TomarIn,
)

router = APIRouter(prefix="/api/orders", tags=["ordenes"])


def _salida(orden: OrdenReparacion, contexto: ApplicationContext) -> OrdenOut:
    """Vista HTTP de la Orden: progreso, acciones e insumos previstos.

    Los insumos previstos se resuelven contra el catalogo al responder.
    Con ellos la UI puede proponer que confirmar en PROC-REP-200 sin
    conocer ningun ID de insumo.
    """
    return OrdenOut.desde_dominio(
        orden,
        progreso=progreso(orden),
        acciones=acciones_disponibles(orden),
        insumos_previstos=insumos_previstos_por_detalle(contexto, orden),
        nombres_de_tipo=nombres_de_tipo_por_detalle(contexto, orden),
    )


def _salida_con_entrega(
    orden: OrdenReparacion,
    contexto: ApplicationContext,
    puede_entregar: bool,
) -> OrdenConEntregaOut:
    """Idem, mas el resultado de PROC-REP-265."""
    return OrdenConEntregaOut(
        **_salida(orden, contexto).model_dump(),
        puede_entregar=puede_entregar,
    )


# --- Lectura -----------------------------------------------------------


@router.get("")
def get_ordenes(contexto: ContextoDep) -> list[OrdenResumenOut]:
    """Listado de Ordenes."""
    return [
        OrdenResumenOut.desde_dominio(orden)
        for orden in listar_ordenes(contexto)
    ]


@router.get("/{orden_id}")
def get_orden(orden_id: str, contexto: ContextoDep) -> OrdenOut:
    """Una Orden con su detalle completo."""
    return _salida(obtener_orden(contexto, orden_id), contexto)


# --- Comandos del Happy Path -------------------------------------------


@router.post("", status_code=status.HTTP_201_CREATED)
def post_crear_orden(
    cuerpo: CrearOrdenIn,
    contexto: ContextoDep,
) -> OrdenOut:
    """Ingreso de un equipo de cliente externo (ACT-RECEP).

    Compone PROC-REP-010 -> 030 -> 040. La Orden nace en REQUERIMIENTO,
    sin Detalles.
    """
    orden = crear_orden(
        contexto,
        usuario_id=cuerpo.usuario_id,
        cliente=cuerpo.cliente.a_dominio(),
        equipo=cuerpo.equipo.a_dominio(),
    )
    return _salida(orden, contexto)


@router.post("/{orden_id}/details")
def post_definir_reparacion(
    orden_id: str,
    cuerpo: DefinirReparacionIn,
    contexto: ContextoDep,
) -> OrdenOut:
    """Definir la reparacion requerida (ACT-RECEP).

    Compone PROC-REP-045 -> 070 -> 050 -> 060 -> 080 -> 090 -> 140:
    despues de definir el Detalle, todo lo que sigue hasta HABILITADA es
    automatico y no cruza otra decision humana. La factibilidad (080)
    consulta el stock global y NO reserva.
    """
    orden = definir_reparacion(
        contexto,
        orden_id=orden_id,
        usuario_id=cuerpo.usuario_id,
        tipo_reparacion_id=cuerpo.tipo_reparacion_id,
        observaciones=cuerpo.observaciones,
    )
    return _salida(orden, contexto)


@router.post("/{orden_id}/queue")
def post_encolar(
    orden_id: str,
    cuerpo: EncolarIn,
    contexto: ContextoDep,
) -> OrdenOut:
    """Priorizar e ingresar a la cola (ACT-COORD).

    Compone PROC-REP-150 -> 170.
    """
    orden = encolar_orden(
        contexto,
        orden_id=orden_id,
        usuario_id=cuerpo.usuario_id,
        prioridad=cuerpo.prioridad,
    )
    return _salida(orden, contexto)


@router.post("/{orden_id}/take")
def post_tomar(
    orden_id: str,
    cuerpo: TomarIn,
    contexto: ContextoDep,
) -> OrdenOut:
    """Tomar la Orden desde una Estacion (ACT-TECH).

    Compone PROC-REP-172 -> 180. Si la validacion de estacion no pasa,
    la aplicacion lo convierte en un conflicto funcional (409) con el
    motivo concreto.
    """
    orden = tomar_orden_en_estacion(
        contexto,
        orden_id=orden_id,
        usuario_id=cuerpo.usuario_id,
        estacion_id=cuerpo.estacion_id,
    )
    return _salida(orden, contexto)


@router.post("/{orden_id}/details/{detalle_id}/start")
def post_iniciar_detalle(
    orden_id: str,
    detalle_id: str,
    cuerpo: IniciarDetalleIn,
    contexto: ContextoDep,
) -> OrdenOut:
    """Iniciar el trabajo sobre un Detalle (ACT-TECH).

    Compone PROC-REP-181 -> 174 -> 185. Aqui ocurre la RESERVA REAL de
    insumos, bajo la seccion critica de inventario.
    """
    orden = iniciar_detalle(
        contexto,
        orden_id=orden_id,
        detalle_id=detalle_id,
        usuario_id=cuerpo.usuario_id,
    )
    return _salida(orden, contexto)


@router.post("/{orden_id}/executions/{ejecucion_id}/complete")
def post_completar_ejecucion(
    orden_id: str,
    ejecucion_id: str,
    cuerpo: CompletarEjecucionIn,
    contexto: ContextoDep,
) -> OrdenOut:
    """Completar la Ejecucion en curso (ACT-TECH propietario).

    Compone PROC-REP-190 -> 200 -> 210 -> 211: el tecnico confirma lo
    realmente utilizado, el inventario se concilia y la situacion de la
    Orden se recalcula.
    """
    orden, _ = completar_ejecucion(
        contexto,
        orden_id=orden_id,
        ejecucion_id=ejecucion_id,
        usuario_id=cuerpo.usuario_id,
        insumos_utilizados=[
            insumo.a_dominio() for insumo in cuerpo.insumos_utilizados
        ],
        observaciones=cuerpo.observaciones,
    )
    return _salida(orden, contexto)


@router.post("/{orden_id}/control/approve")
def post_aprobar_control(
    orden_id: str,
    cuerpo: AprobarControlIn,
    contexto: ContextoDep,
) -> OrdenOut:
    """Aprobar el control tecnico (ACT-RECEP).

    Compone PROC-REP-220 -> 230 -> 245 -> 240. No notifica: avisar al
    cliente es otra decision de Recepcion.
    """
    orden = aprobar_control(
        contexto,
        orden_id=orden_id,
        usuario_id=cuerpo.usuario_id,
        observaciones=cuerpo.observaciones,
    )
    return _salida(orden, contexto)


@router.post("/{orden_id}/notify")
def post_notificar(
    orden_id: str,
    cuerpo: NotificarIn,
    contexto: ContextoDep,
) -> OrdenConEntregaOut:
    """Notificar al cliente y evaluar el cobro (ACT-RECEP).

    Compone PROC-REP-250 -> 260 -> 265, y PROC-REP-266 si queda saldo.
    """
    orden, puede_entregar = notificar(
        contexto,
        orden_id=orden_id,
        usuario_id=cuerpo.usuario_id,
    )
    return _salida_con_entrega(orden, contexto, puede_entregar)


@router.post("/{orden_id}/payments")
def post_registrar_pago(
    orden_id: str,
    cuerpo: PagoIn,
    contexto: ContextoDep,
) -> OrdenConEntregaOut:
    """Registrar un Pago (capacidad transversal).

    FEAT-REP-007 / BR-REP-017. No corresponde a ningun PROC-REP-* propio
    y no exige un rol: el negocio todavia no definio cual. Despues del
    pago se revalida PROC-REP-265.
    """
    orden, puede_entregar = registrar_pago_de_orden(
        contexto,
        orden_id=orden_id,
        usuario_id=cuerpo.usuario_id,
        monto=cuerpo.monto,
        metodo=cuerpo.metodo,
    )
    return _salida_con_entrega(orden, contexto, puede_entregar)


@router.post("/{orden_id}/deliver")
def post_entregar(
    orden_id: str,
    cuerpo: EntregarIn,
    contexto: ContextoDep,
) -> OrdenOut:
    """Entregar el equipo (ACT-ADMIN).

    Compone PROC-REP-280 -> 270 -> EVT-REP-999. La Orden queda ENTREGADA.
    """
    orden = entregar(
        contexto,
        orden_id=orden_id,
        usuario_id=cuerpo.usuario_id,
    )
    return _salida(orden, contexto)

File: app/frontend/src/api/ordenes.ts
/**
 * Llamadas a la API de Ordenes y catalogos.
 *
 * Una funcion por intencion de usuario, con los mismos nombres que los
 * comandos del backend. Cada POST devuelve la Orden ya actualizada, asi
 * que la UI nunca tiene que recomponer el estado por su cuenta.
 */

import type {
  Estacion,
  InsumoUtilizado,
  Orden,
  OrdenResumen,
  TipoReparacion,
  Usuario,
} from "../types/api";
import { get, post } from "./client";

// --- Catalogos ---------------------------------------------------------

export function listarTiposReparacion(): Promise<TipoReparacion[]> {
  return get<TipoReparacion[]>("/api/catalogs/repair-types");
}

export function listarEstaciones(): Promise<Estacion[]> {
  return get<Estacion[]>("/api/catalogs/stations");
}

export function listarUsuarios(): Promise<Usuario[]> {
  return get<Usuario[]>("/api/catalogs/users");
}

// --- Lectura de Ordenes ------------------------------------------------

export function listarOrdenes(): Promise<OrdenResumen[]> {
  return get<OrdenResumen[]>("/api/orders");
}

export function obtenerOrden(ordenId: string): Promise<Orden> {
  return get<Orden>(`/api/orders/${ordenId}`);
}

// --- Comandos del Happy Path -------------------------------------------

export interface DatosNuevaOrden {
  usuario_id: string;
  cliente: { nombre: string; telefono: string; email?: string | null };
  equipo: { marca: string; modelo: string; falla_reportada: string };
}

/** PROC-REP-010 -> 030 -> 040 (Recepción). */
export function crearOrden(datos: DatosNuevaOrden): Promise<Orden> {
  return post<Orden>("/api/orders", datos);
}

/** PROC-REP-045 -> 070 -> 050 -> 060 -> 080 -> 090 -> 140 (Recepción). */
export function definirReparacion(
  ordenId: string,
  usuarioId: string,
  tipoReparacionId: string,
): Promise<Orden> {
  return post<Orden>(`/api/orders/${ordenId}/details`, {
    usuario_id: usuarioId,
    tipo_reparacion_id: tipoReparacionId,
  });
}

/** PROC-REP-150 -> 170 (Coordinador). */
export function encolar(
  ordenId: string,
  usuarioId: string,
  prioridad: number,
): Promise<Orden> {
  return post<Orden>(`/api/orders/${ordenId}/queue`, {
    usuario_id: usuarioId,
    prioridad,
  });
}

/** PROC-REP-172 -> 180 (Técnico). */
export function tomarOrden(
  ordenId: string,
  usuarioId: string,
  estacionId: string,
): Promise<Orden> {
  return post<Orden>(`/api/orders/${ordenId}/take`, {
    usuario_id: usuarioId,
    estacion_id: estacionId,
  });
}

/** PROC-REP-181 -> 174 -> 185: reserva real (Técnico). */
export function iniciarDetalle(
  ordenId: string,
  detalleId: string,
  usuarioId: string,
): Promise<Orden> {
  return post<Orden>(`/api/orders/${ordenId}/details/${detalleId}/start`, {
    usuario_id: usuarioId,
  });
}

/** PROC-REP-190 -> 200 -> 210 -> 211 (Técnico propietario). */
export function completarEjecucion(
  ordenId: string,
  ejecucionId: string,
  usuarioId: string,
  insumosUtilizados: InsumoUtilizado[],
  observaciones: string | null,
): Promise<Orden> {
  return post<Orden>(
    `/api/orders/${ordenId}/executions/${ejecucionId}/complete`,
    {
      usuario_id: usuarioId,
      insumos_utilizados: insumosUtilizados,
      observaciones,
    },
  );
}

/** PROC-REP-220 -> 230 -> 245 -> 240 (Recepción). */
export function aprobarControl(
  ordenId: string,
  usuarioId: string,
  observaciones: string | null,
): Promise<Orden> {
  return post<Orden>(`/api/orders/${ordenId}/control/approve`, {
    usuario_id: usuarioId,
    observaciones,
  });
}

/** PROC-REP-250 -> 260 -> 265 [-> 266] (Recepción). */
export function notificar(
  ordenId: string,
  usuarioId: string,
): Promise<Orden> {
  return post<Orden>(`/api/orders/${ordenId}/notify`, {
    usuario_id: usuarioId,
  });
}

/** Capacidad transversal FEAT-REP-007 / BR-REP-017, sin rol exigido. */
export function registrarPago(
  ordenId: string,
  usuarioId: string,
  monto: string,
  metodo: string,
): Promise<Orden> {
  return post<Orden>(`/api/orders/${ordenId}/payments`, {
    usuario_id: usuarioId,
    monto,
    metodo,
  });
}

/** PROC-REP-280 -> 270 -> EVT-REP-999 (Administrador). */
export function entregar(
  ordenId: string,
  usuarioId: string,
): Promise<Orden> {
  return post<Orden>(`/api/orders/${ordenId}/deliver`, {
    usuario_id: usuarioId,
  });
}
